cassandra/src/main/scala/zio/flow/internal/CassandraKeyValueStore.scala

```diff
@@ -0,0 +1,219 @@
+
+  override def put(
+    namespace: String,
+    key: Chunk[Byte],
+    value: Chunk[Byte]
+  ): IO[IOException, Boolean] = {
+    val insert =
+      QueryBuilder
+        .insertInto(keyspace, tableOf(namespace))
+        .value(
+          keyColumnName,
+          byteBufferFrom(key)
+        )
+        .value(
+          valueColumnName,
+          byteBufferFrom(value)
+        )
+        .build
+
+    val operation =
+      executeAsync(insert, session).as(true)
+
+    operation.catchSome {
+      case error: InvalidQueryException if tableDoesNotExist(error, namespace) =>
+        val createTable =
+          SchemaBuilder
+            .createTable(keyspace, tableOf(namespace))
+            .withPartitionKey(keyColumnName, DataTypes.BLOB)
+            .withColumn(valueColumnName, DataTypes.BLOB)
+            .build
+
+        executeAsync(createTable, session) *> operation
+    }.mapError(
+      refineToIOException("Error putting key-value pair")
+    )
+  }
+
+  override def get(namespace: String, key: Chunk[Byte]): IO[IOException, Option[Chunk[Byte]]] = {
+    val query =
+      QueryBuilder
+        .selectFrom(keyspace, tableOf(namespace))
+        .column(valueColumnName)
+        .whereColumn(keyColumnName)
+        .isEqualTo(
+          byteBufferFrom(key)
+        )
+        .limit(1)
+        .build
+
+    val operation =
+      executeAsync(query, session).map { result =>
+        if (result.remaining > 0)
+          Option(
+            blobValueOf(valueColumnName, result.one)
+          )
+        else None
+      }.catchSome {
+        case error: InvalidQueryException if tableDoesNotExist(error, namespace) =>
+          ZIO.none
+      }
+
+    operation.mapError(
+      refineToIOException("Error retrieving or reading value")
+    )
+  }
+
+  override def scanAll(namespace: String): ZStream[Any, IOException, (Chunk[Byte], Chunk[Byte])] = {
+    val operation =
+      ZStream.unwrap {
+        val query =
+          QueryBuilder
+            .selectFrom(keyspace, tableOf(namespace))
+            .column(keyColumnName)
+            .column(valueColumnName)
+            .build
+
+        executeAsync(query, session).map { pageResult1 =>
+          val morePages =
+            if (pageResult1.hasMorePages)
+              pageMoreFrom(pageResult1)
+            else
+              ZStream.empty
+
+          val retrieved = byteChunkPairsFrom(pageResult1)
+
+          ZStream
+            .fromChunk(retrieved)
+            .concat(morePages)
+        }
+      }.catchSome {
+        case error: InvalidQueryException if tableDoesNotExist(error, namespace) =>
+          ZStream.empty
+      }
+
+    operation.mapError(
+      refineToIOException("Error scanning all key-value pairs")
+    )
+  }
+
+  private def executeAsync(statement: SimpleStatement, session: CqlSession) =
+    ZIO.fromCompletionStage(
+      session.executeAsync(statement)
+    )
+
+  private def pageMoreFrom(pageResult1: AsyncResultSet) =
```

Comment:
You may be able to use `ZStream.paginate` here.

Reply:
Ah yes, `ZStream.paginate`. I actually used it before elsewhere (for paging ElasticSearch results) but I forgot about it :) Will try replacing this with `paginate`.

Currently this `pageMoreFrom` is inspired by a method in vigoo's zio-aws (https://github.com/vigoo/zio-aws/blob/master/zio-aws-core/src/main/scala/io/github/vigoo/zioaws/core/AwsServiceBase.scala#L45)

Reply:
Implementation is using `paginate` now.